Bugzilla/Net/Storage.pm

```diff
@@ -0,0 +1,200 @@
+has error_string => (is => 'rw',   isa      => Str);
+has port         => (is => 'ro',   isa      => Int);
+
+##################
+# Public Methods #
+##################
+
+# Add new data to net storage
+sub add_key {
+  my ($self, $key, $value) = @_;
+
+  ThrowCodeError('net_storage_invalid_key')   unless $key   && length $key;
+  ThrowCodeError('net_storage_invalid_value') unless $value && length $value;
+
+  my ($method, $path) = $self->_get_method_path('add', $key);
+
+  return $self->_send_request($method, $path, $value);
+}
+
+# Check if a key exists in net storage
+sub head_key {
+  my ($self, $key) = @_;
+
+  ThrowCodeError('net_storage_invalid_key') unless $key && length $key;
+
+  my ($method, $path) = $self->_get_method_path('head', $key);
+
+  return $self->_send_request($method, $path);
+}
+
+# Get data from net storage
+sub get_key {
+  my ($self, $key) = @_;
+
+  ThrowCodeError('net_storage_invalid_key') unless $key && length $key;
+
+  my ($method, $path) = $self->_get_method_path('get', $key);
+
+  return $self->_send_request($method, $path);
+}
+
+sub delete_key {
+  my ($self, $key) = @_;
+
+  ThrowCodeError('net_storage_invalid_key') unless $key && length $key;
+
+  my ($method, $path) = $self->_get_method_path('delete', $key);
+
+  return $self->_send_request($method, $path);
+}
+
+###################
+# Private Methods #
+###################
+
+sub _build_ua {
+  my ($self) = @_;
+
+  my $ua;
+  if ($self->retry) {
+    require LWP::UserAgent::Determined;
+    $ua = LWP::UserAgent::Determined->new(
+      agent                 => 'Bugzilla',
+      keep_alive            => 10,
+      requests_redirectable => [qw(GET HEAD DELETE PUT)],
+    );
+    $ua->timing('1,2,4,8,16,32');
+  }
+  else {
+    $ua = LWP::UserAgent->new(
+      agent                 => 'Bugzilla',
+      keep_alive            => 10,
+      requests_redirectable => [qw(GET HEAD DELETE PUT)],
+    );
+  }
+  $ua->timeout($self->timeout);
+  if (my $proxy = Bugzilla->params->{proxy_url}) {
+    $ua->proxy(['https', 'http'], $proxy);
+  }
+
+  return $ua;
+}
+
+sub _remember_errors {
+  my ($self, $response) = @_;
+  $self->error_code($response->code);
+  $self->error_string($response->status_line);
+}
+
+# make the HTTP::Request object
+sub _make_request {
+  my ($self, $method, $path, $data, $headers) = @_;
+  $headers ||= {};
+  $data //= '';
+
+  my $http_headers = HTTP::Headers->new;
+  foreach my $key (keys %{$headers}) {
+    $http_headers->header($key => $headers->{$key});
+  }
+
+  $self->_add_auth_header($http_headers, $method, $path);
+
+  my $protocol = $self->secure ? 'https'                         : 'http';
+  my $host     = $self->port   ? $self->host . ':' . $self->port : $self->host;
+
+  # Update some settings if we are running in a test environment
+  if ($self->host eq 's3') {
+    $protocol = 'http';
+    $host     = 's3:9000';
+  }
+  elsif ($self->host eq 'gcs') {
+    $protocol = 'http';
+    $host     = 'gcs:4443';
+  }
```

Comment:
These pieces should be in the subclasses with a common method that is overridden.